### ISSUE TITLE
Qt apps not completely following themes

### ISSUE THREAD

**Issue by lallenfrancisl** [flatpak/flatpak#2030]
## Desktop Environment: Budgie
The app I am using to show this problem is vlc, but it happens in other apps too. Although the theme(I am currently using adapta) is downloaded with the runtime its not applied consistently across every element. Like the menubar in vlc. 
![screenshot from 2018-08-28 11-54-05](https://user-images.githubusercontent.com/7833263/45057738-4b782180-b04b-11e8-86c8-47799bb6469e.png)
Just installing vlc or sublime(maybe some other apps) is sufficient for reproducing the issue

**Comment by TingPing** [comment]
So assuming that this is the [adapta-kde](https://github.com/PapirusDevelopmentTeam/adapta-kde) theme it looks like this needs to be done to work:

- [Kvantum](https://github.com/tsujan/Kvantum) needs to be packaged for the kde runtime
- Kvantum based themes like Adapta then have to be packaged for that
- Then we need some way for Kvantum to actually figure out which theme is set because it loads files from `~/.config` on the host which is broken in a sandbox

Maybe KDE folks have more insight but I don't think this is simple to solve.

**Comment by TingPing** [comment]
Oh sorry I missed you mentioned Sublime. That uses Gtk2 which is a different problem, arguably easier to solve but not planned on happening at the moment with the hope software moves to Gtk3.

**Comment by lallenfrancisl** [comment]
Oh thanks @TingPing . But small things like these are going to actually prevent people from using. Just my opinion. Should I close this issue then.

**Comment by TingPing** [comment]
I do care about the theming situation but Qt is not an easy problem to solve.

**Comment by lallenfrancisl** [comment]
Should I close this issue @TingPing or should I not then?

**Comment by lallenfrancisl** [comment]
Actually @TingPing  I am not using adapta-kde I am using this [adapta-gtk](https://github.com/adapta-project/adapta-gtk-theme). I think you mean the wrong one.

**Comment by TingPing** [comment]
If Qt is loading the Gtk2 theme that might be easier but still problematic.

**Comment by lallenfrancisl** [comment]
@TingPing  I get you. You are saying that qt only supports qt themes, and me using gtk themes does not load up. right?

**Comment by TingPing** [comment]
I'm saying Qt handles themes its own way that is hard to Flatpak.

**Comment by lallenfrancisl** [comment]
@TingPing I am also seeing this in Visual studio code

**Comment by TiZ-EX1** [comment]
Hi, I'm curious about this too. I would like to use flatpak versions of VLC and other Qt apps. I've installed `org.kde.PlatformTheme.QGnomePlatform`, but it seems to just use Adwaita instead of the Adapta I've installed in `~/.themes`. Is QGnomePlatform different from QGtkStyle? If so, is there a reason QGtkStyle is not packaged aside from just someone hasn't had incentive to do it?

**Comment by TingPing** [comment]
> Is QGnomePlatform different from QGtkStyle? If so, is there a reason QGtkStyle is not packaged aside from just someone hasn't had incentive to do it?

Well QGtkStyle is forever stuck in the GTK2 past and none of those themes are packaged anywhere.

**Comment by TiZ-EX1** [comment]
None of the GTK2 engines are packaged, right? I remember us having this conversation. So even if QGtkStyle were packaged just to load up stuff in ~/.themes, none of it would look right, correct? Presumably the proper way forward is installing the Adapta Qt theme.

**Comment by TingPing** [comment]
GTK2 is no longer packaged in any runtime and loading files in `~` will usually fail.

**Comment by ilmaisin** [comment]
Is font size dependent on this or something else? I run Centos 7 and Qt apps installed via flatpak/Flathub seem to ignore font scaling setting.

**Comment by Esokrates** [comment]
Where should I open a report requesting the packaging of Kvantum?
Adwaita-qt ist not very actively maintained, see e.g. https://github.com/FedoraQt/adwaita-qt/issues/110 it does currently not even match the current Adwaita theme.
Apart from that adwaita-qt has a few issues, like scrollbars not matching or details like https://github.com/FedoraQt/adwaita-qt/issues/54.
All of these are solved in Kvantum, the developer is very passionate about his work and even promised to update the matching KvGnome theme every half year matching Gnome's release cylce, see https://github.com/tsujan/Kvantum/issues/347

So even for a uniform Adwaita look it would be great to package Kvantum. Furthermore Kvantum provides many themes matching popular gtk themes, e.g. the Ubuntu Yaru theme is a perfect match too, so making Kvantum work in Flatpak would make theming so much more perfect.

**Comment by Lonsfor** [comment]
i know there is already a pr for Kvantum but i don't know in what state it is in
https://github.com/flathub/flathub/pull/779

**Comment by gasinvein** [comment]
@TingPing
> * [Kvantum](https://github.com/tsujan/Kvantum) needs to be packaged for the kde runtime

Yeah, here flathub/flathub#779.

> * Kvantum based themes like Adapta then have to be packaged for that

Not nessesary, Kvantum provides a default set of most popular themes, Adapta included.

> * Then we need some way for Kvantum to actually figure out which theme is set because it loads files from `~/.config` on the host which is broken in a sandbox

Kvantum can automantically apply theme matching current GTK theme.

**Comment by TingPing** [comment]
@gasinvein Users will expect existing config files to work and they won't. But if it can work without a sandbox escape thats cool and just needs to be documented.

**Comment by gasinvein** [comment]
@TingPing it requires additional action, i.e. `flatpak override --fileystsem=xdg-config/Kvantum:ro`. Is this considered a sandbox escape?

**Comment by TingPing** [comment]
Escape is the wrong word but any permission is not ideal and broken by default.

If Kvantum themes are also anything more than plain-text configuration they are probably broken in general.

**Comment by gasinvein** [comment]
Kvantum themes are ini-like configuration + svg image.

**Comment by memeplex** [comment]
I've been following this thread because I use some qt apps with kvantum+materia inside a gnome environment. Currently I'm using arch but I'm considering moving to fedora so reliance in flatpak would be an important aspect of my decision. I believe flatpak (and snaps) are the way forward for app distribution but I'm prudent and currently having rolling releases and the aur gives me a lot of flexibility (not without costs, of course). So I have this question: even if you packaged kvantum as part of a kde runtime, how would I configure the specific theme kvantum will be using? It's not like kvantum automagically detects the current gtk theme and chooses the one that resembles it the most. Thank you  in advance for any explanation on this topic.

**Comment by TingPing** [comment]
The runtime would add a permission to the kvantum config dir. According to the info posted here it should work fine.

**Comment by memeplex** [comment]
So the idea is that the user manually edits those files or that he/she launches the kvantum manager gui from flatpak?

**Comment by TingPing** [comment]
How they currently configure kvantum can probably keep working as-is.

**Comment by memeplex** [comment]
It's just a matter of launching kvantum manager and changing the theme from there. But there will be a couple of managers in the system, the standard one and the flatpak one. I'm not sure how that will work.

**Comment by TingPing** [comment]
I'm not sure what you are talking about honestly. As discussed here kvantum has a single configuration location. A single tool can configure said single location. If it references system installed files then it will have to be packaged but otherwise nothing is different.

**Comment by memeplex** [comment]
Ok, let's see then. I will have at least twice kvantum installations AFAICS: the usual one obtained from my distro package manager and the one installed in a flatpak runtime. Each one will have its own set of configuration files. So far so good. Now, these files are not usually edited by hand but from a kvantum manager gui that is an integral part of the kvantum package. Indeed, the dotfiles can easily become non-trivial if you tweak a bit the myriad of kvantum options. So, again AFAICS, there will be two manager guis installed in my system. I assume that, for example, when launching them from gnome shell, two launchers will be shown in the activities overview, which can be quite confusing.

**Comment by TiZ-EX1** [comment]
Hold on. I think I get what you're not getting.

The flatpak version of kvantum and the system version of kvantum **will not** have differing sets of configurations. When it comes to GTK, both the system and flatpak versions read from the user configurations, wherever those happen to be (xfconf, dconf, $XDG_CONFIG_HOME/gtk-3.0, wherever). Permissions are given to allow applications using GTK to do that.

There is no reason kvantum should be any different. The main things that are duplicated between a system kvantum package and a flatpak kvantum package are just the libraries, allowing the KDE/Qt framework and apps using them to utilize kvantum, and then peek outside the sandbox with the window we provide to see the user configuration that already exists.

tl;dr: system kvantum config tool will work fine for flatpak kvantum.

**Comment by memeplex** [comment]
Oh, that's great, no duplication at all. I was also investigating about the gtk mechanism to achieve the "theme sync" and you even clarified that. Thanks!

> wherever those happen to be (xfconf, dconf, $XDG_CONFIG_HOME/gtk-3.0,

How is this achieved? I mean, gtk has a complex lookup sequence that would require to export a lot of places in the filesystem hierarchy to the sandbox. How is the configuration in the sandbox merged with the user configuration outside the sandbox? And does this also work for, say, a font antialias setting in my dconf? (sorry for the many questions, but knowing that flatpak can access my user configuration is a good notice and now I want to understand a bit more).

**Comment by memeplex** [comment]
I mean, it's obvious that exporting the entire home directory would solve it, but it's also obvious that there are security concerns and potential changes of configuration interfaces in doing so.

**Comment by mbnoimi** [comment]
Is there any progress on this bug? 
I still able to reproduce it on my Clementine Player
* OS: Linux Mint 20 xfce
* Theme: Mint-Y-Dark-Aqua

Default package manager screenshot (apt)
![image](https://user-images.githubusercontent.com/284547/100973845-64cee780-354c-11eb-9639-a8110167505c.png)

Flatpak 1.9.2 screenshot
![image](https://user-images.githubusercontent.com/284547/100974096-c42cf780-354c-11eb-8785-6f4f00145e75.png)

This didn't fix the issue!
```
$ flatpak install flathub org.gtk.Gtk3theme.Mint-Y-Dark-Aqua
Looking for matches…
Skipping: org.gtk.Gtk3theme.Mint-Y-Dark-Aqua/x86_64/3.22 is already installed
```

**Comment by HugLifeTiZ** [comment]
Neither Kvantum or QGtkStyle are packaged in Flathub. Qt apps in Flathub are simply not going to follow your GTK theme.

**Comment by nigelwarning** [comment]
Is this a Qt or Flatpak issue? On KDE self Qt flatpaks works fine with the system theme. On every other distro not. While setting a GTK theme on KDE for GTK Flatpaks is really easy

**Comment by hfiguiere** [comment]
This is mostly a runtime problem, and a toolkit problem.
